File: src/main/java/no/nav/syfo/domain/dialogmelding/RSHodemelding.java
package no.nav.syfo.domain.dialogmelding;

import lombok.Getter;
import no.nav.syfo.domain.*;
import no.nav.syfo.domain.oppfolgingsplan.RSOppfolgingsplan;

import static java.util.Optional.ofNullable;

@Getter
public class RSHodemelding {
    private RSMeldingInfo meldingInfo;
    private RSVedlegg vedlegg;

    public RSHodemelding(Fastlege fastlege, Partnerinformasjon partnerinformasjon, RSOppfolgingsplan oppfolgingsplan){
        this.meldingInfo =  tilMeldingInfo(
                tilMottaker(fastlege, partnerinformasjon),
                tilPasient(fastlege.pasient()));
        this.vedlegg = tilVedlegg(oppfolgingsplan);
    }

    private RSMeldingInfo tilMeldingInfo(RSMottaker mottaker, RSPasient pasient) {
        return new RSMeldingInfo(mottaker, pasient);
    }

    private RSPasient tilPasient(Pasient pasient) {
        return new RSPasient(
                pasient.fnr(),
                pasient.fornavn(),
                pasient.mellomnavn(),
                pasient.etternavn());
    }

    private RSVedlegg tilVedlegg(RSOppfolgingsplan oppfolgingsplan) {
        return new RSVedlegg(oppfolgingsplan.getOppfolgingsplanPdf());
    }

    private RSBehandler tilBehandler(Fastlege fastlege) {
        return new RSBehandler(
                fastlege.fnr(),
                fastlege.helsepersonellregisterId(),
                fastlege.fornavn(),
                fastlege.mellomnavn(),
                fastlege.etternavn());
    }

    private RSMottaker tilMottaker(Fastlege fastlege, Partnerinformasjon partnerinformasjon) {
        return new RSMottaker(
                partnerinformasjon.getPartnerId(),
                partnerinformasjon.getHerId(),
                fastlege.fastlegekontor().orgnummer(),
                fastlege.fastlegekontor().navn(),
                ofNullable(fastlege.fastlegekontor().postadresse()).map(Adresse::adresse).orElse(null),
                ofNullable(fastlege.fastlegekontor().postadresse()).map(Adresse::postnummer).orElse(null),
                ofNullable(fastlege.fastlegekontor().postadresse()).map(Adresse::poststed).orElse(null),
                tilBehandler(fastlege));
    }
}

File: src/main/kotlin/no/nav/syfo/dialogmelding/domain/RSHodemelding.kt
package no.nav.syfo.dialogmelding.domain

import no.nav.syfo.dialogmelding.api.RSOppfolgingsplan
import no.nav.syfo.fastlege.domain.*

data class RSHodemelding(
    val meldingInfo: RSMeldingInfo,
    val vedlegg: RSVedlegg,
)

fun createRSHodemelding(
    fastlege: Fastlege,
    partnerinformasjon: Partnerinformasjon,
    oppfolgingsplan: RSOppfolgingsplan,
) = RSHodemelding(
    meldingInfo = tilMeldingInfo(
        mottaker = fastlege.tilMottaker(partnerinformasjon),
        pasient = fastlege.pasient.tilPasient(),
    ),
    vedlegg = oppfolgingsplan.tilVedlegg(),
)

private fun tilMeldingInfo(
    mottaker: RSMottaker,
    pasient: RSPasient,
) = RSMeldingInfo(
    mottaker = mottaker,
    pasient = pasient,
)

private fun Pasient?.tilPasient() = RSPasient(
    fnr = this?.fnr,
    fornavn = this?.fornavn,
    mellomnavn = this?.mellomnavn,
    etternavn = this?.etternavn,
)

private fun RSOppfolgingsplan.tilVedlegg() = RSVedlegg(this.oppfolgingsplanPdf)

private fun Fastlege.tilMottaker(
    partnerinformasjon: Partnerinformasjon,
) = RSMottaker(
    partnerId = partnerinformasjon.partnerId.toString(),
    herId = partnerinformasjon.herId.toString(),
    orgnummer = this.fastlegekontor.orgnummer,
    navn = this.fastlegekontor.navn,
    adresse = this.fastlegekontor.postadresse?.adresse,
    postnummer = this.fastlegekontor.postadresse?.postnummer,
    poststed = this.fastlegekontor.postadresse?.poststed,
    behandler = this.tilBehandler(),
)

private fun Fastlege.tilBehandler() = RSBehandler(
    this.fnr,
    this.helsepersonellregisterId,
    this.fornavn,
    this.mellomnavn,
    this.etternavn,
)
